workflow/workflow.go

```diff
@@ -333,19 +335,23 @@ func (l *loopState) notifySteps() { //nolint:gocognit
 	// concrete values that needed to be updated. This would make it possible to completely forego the need to
 	// iterate through the input.
 
-	nodesWithoutInbound := l.dag.ListNodesWithoutInboundConnections()
-	l.logger.Debugf("Currently %d DAG nodes have no inbound connection. Now processing them.", len(nodesWithoutInbound))
+	readyNodes := l.dag.PopReadyNodes()
+	l.logger.Debugf("Currently %d DAG nodes are ready. Now processing them.", len(readyNodes))
 
-	// nodesWithoutInbound have all dependencies resolved. No inbound connection.
-	// Also includes nodes that are not for running, like an input.
-	for nodeID, node := range nodesWithoutInbound {
-		if _, ok := l.inputsNotified[nodeID]; ok {
+	// Can include runnable nodes, nodes that cannot be resolved, and nodes that are not for running, like an input.
+	for nodeID, resolutionStatus := range readyNodes {
+		if resolutionStatus == dgraph.Unresolvable {
+			l.logger.Debugf("Disregarding node %q with resolution %q", nodeID, resolutionStatus)
 			continue
 		}
 		l.logger.Debugf("Processing step node %s", nodeID)
-		l.inputsNotified[nodeID] = struct{}{}
+		node, err := l.dag.GetNodeByID(nodeID)
+		if err != nil {
+			panic(fmt.Errorf("failed to get node %s (%w)", nodeID, err))
+		}
+		nodeItem := node.Item()
 		// The data structure that the particular node requires. One or more fields. May or may not contain expressions.
-		inputData := node.Item().Data
+		inputData := nodeItem.Data
 		if inputData == nil {
 			// No input data is needed.
 			continue
```

Comment:
I don't think that's obvious without the comment.

Reply:
Without the comment, it's pretty clear that the code does nothing but skip over this node (i.e., it's clear that no action is needed for this node).  What is unclear, perhaps, is _why_ the code is skipping this node.  I believe that the reason for this is that this node represents input data and not an executable step, but the comment doesn't say anything like that (to me).

Contrary to your assertion, the comment, "No input data is needed" does not suggest to me that this node **is** a non-executable _input_ node.

Reply:
I added to the comment.

Reply:
Adding to the comment doesn't really help much, because the part that I'm objecting to is the part that you didn't change:  I don't understand what you think "input data" and "needed" mean in this context.  The question which this comment should answer is, "Why are we returning at this point?" or "What does an `inputData` value of `nil` mean?"  And, the answer is not that the unnamed something doesn't need input data -- the answer is that the node from which we pulled `inputData` itself has no inputs because it _is_ the **input** to the workflow; it is non-executable, existing only to satisfy the dependencies of other nodes, and _therefore_ we skip it and just continue with the next node.

Vague statements like "This is often the case" and a reference to "input nodes" without further explanation just isn't helpful.  (There's only ever _one_ input node, right?  And, it is totally unlike any other node, because its `node.Item().Data` is `nil`, right?)

So, I recommend a comment along the lines of
> This is the workflow input node; skip it because it doesn't require any processing.

What is tragic is, if we follow my suggestion above to remove this node from the ready list when we resolve it, then we don't need this code here and so we don't need to argue over how to explain what it does.  😞